ai-24sp/prototypes/prototype-00/network.py

```diff
@@ -1,46 +1,178 @@
-    gradient descent. The "training data" is a list of tuples
-    "(x, y)" representing the training inputs and the desired
-    outputs. The other non-optional parameters are
-    self-explanatory. If "test_data" is provided then the
-    network will be evaluated against the test data after each
-    epoch, and partial progress printed out. This is useful for
-    tracking progress, but slows things down substantially."""
-    if test_data:
-        n_test = len(test_data)
-    n = len(training_data)
-
-    for j in xrange(epochs):
-        random.shuffle(training_data)
-        mini_batches = [
-            training_data[k:k+mini_batch_size]
-            for k in xrange(0, n, mini_batch_size)
-        ]
-
-        for mini_batch in mini_batches:
-            self.update_mini_batch(mini_batches, eta)
\ No newline at end of file
+            z = np.dot(w, activation)+b
+            zs.append(z)
+            activation = sigmoid(z)
+            activations.append(activation)
+        # backward pass
+        delta = self.cost_derivative(activations[-1], y) * \
+            sigmoid_prime(zs[-1])
+        nabla_b[-1] = delta
+        nabla_w[-1] = np.dot(delta, activations[-2].transpose())
+        # Note that the variable l in the loop below is used a little
+        # differently to the notation in Chapter 2 of the book.  Here,
+        # l = 1 means the last layer of neurons, l = 2 is the
+        # second-last layer, and so on.  It's a renumbering of the
+        # scheme in the book, used here to take advantage of the fact
+        # that Python can use negative indices in lists.
+        for l in range(2, self.num_layers):
+            z = zs[-l]
+            sp = sigmoid_prime(z)
+            delta = np.dot(self.weights[-l+1].transpose(), delta) * sp
+            nabla_b[-l] = delta
+            nabla_w[-l] = np.dot(delta, activations[-l-1].transpose())
+        return (nabla_b, nabla_w)
+
+    def evaluate(self, test_data):
+        """Return the number of test inputs for which the neural
+        network outputs the correct result. Note that the neural
+        network's output is assumed to be the index of whichever
+        neuron in the final layer has the highest activation."""
+        test_results = [(np.argmax(self.feedforward(x)), y)
+                        for (x, y) in test_data]
+        return sum(int(x == y) for (x, y) in test_results)
+
+    def single(self, test_image):
+        """
+        Return the output classification label for a single input image.
+        Useful for validating or testing a loaded model.
+        """
+        return np.argmax(self.feedforward(test_image))
+
+    def cost_derivative(self, output_activations, y):
+        """Return the vector of partial derivatives \partial C_x /
+        \partial a for the output activations."""
+        return (output_activations-y)
+
+#### Miscellaneous functions
+def sigmoid(z):
+    """The sigmoid function."""
+    return 1.0/(1.0+np.exp(-z))
+
```

Comment:
I believe this sigmoid function will fail with integer overflow on large negative inputs. It can be solved by using a different form of the sigmoid function when `z` is negative
```
def sigmoid(z):
    output = 0
    if z < 0:
        output = np.exp(z) /(1+np.exp(z))
    else:
        output = 1.0/(1.0+np.exp(-z))
    return output
```
or is this the same as what you have defined as `sigmoid_prime`? either way I get an integer overflow error when running `train.py`.

Reply:
`sigmoid` is normally called with an array / vector, so the comparison `z < 0` won't work. Let's keep thinking about what the vectorized equivalent would be.